### PR TITLE
Introduce a new platform layer type for Separated Images

### PR DESCRIPTION
**Pull request by etiennesegonzac**
#### 8edd622a31cf499a31cfb2958ae7f23aeca8a5ad
<pre>
Introduce a new platform layer type for Separated Images
<a href="https://bugs.webkit.org/show_bug.cgi?id=283721">https://bugs.webkit.org/show_bug.cgi?id=283721</a>
&lt;<a href="https://rdar.apple.com/137133500">rdar://137133500</a>&gt;

Reviewed by Simon Fraser.

Initially keyed on the `separated` transform style, and behind a
configuration flag, this patch introduces a new platformLayer Type
/ Custom View for Separated Images.

This view uses takes the layer&apos;s backing store IOSurface. And this
behavior is currently only enabled for images below the dimension
thresholds triggering tiling.

* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::GraphicsLayer):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::isSeparatedImage const):
(WebCore::GraphicsLayer::setIsSeparatedImage):
New method to flag a graphics layer as a Separated Image.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setIsSeparatedImage):
Flag the graphics layer as a SepratedImage, will impact layer type.
(WebCore::GraphicsLayerCA::commitLayerTypeChangesBeforeSublayers):
Change the platform layer type for separated images.
(WebCore::GraphicsLayerCA::changeLayerTypeTo):
Make sure Separated flag and clipping strategy are not lost when
changing layer type.

* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::PlatformCALayerCocoa):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
Introduce new platform Layer Type for Separated Images.

* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateSeparatedProperties):
Introduce a new method to update properties related to the Separated
transform style.
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateGeometry):
(WebCore::RenderLayerBacking::contentChanged):
Update Separated properties as part of the configuration update or when
the Image content changes.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerStyleChanged):
Flag the layer as needing a compositing configuration change when the
Separated transform-style is updated.

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
WKA entry for the WKSeparatedImageView.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
Add a custom backing store application method for WKSeparatedImageViews,
passing the IOSurface to the view.

* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.h:
* Source/WebKit/Shared/API/Cocoa/WKRemoteLayerTreeViewProtocols.h: Added.
Expose the WKContentControlled protocol to Swift.

* Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.h: Added.
* Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.mm: Added.
(-[WKSeparatedImageView init]):
(-[WKSeparatedImageView initWithFrame:]):
(-[WKSeparatedImageView initWithCoder:]):
(-[WKSeparatedImageView setSurface:]):
Stub implementation for public SDK builds.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Instantiate a custom view, WKSeparatedImageView for the Separated Image
Layer Type.

* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(allowListedClassToString):
(shouldDumpSeparatedDetails):
(dumpUIView):
Adjust the content of the text dump to get a reliable tests without
dynamic IDs. Dump layer corner radius for UIViews too.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-content-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-content.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-tiled-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-tiled.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-tiled-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-tiled.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-expected.txt: Added.
Add new UIView-based tests covering upgrading to and downgrading from a
Separated Image platform layer based on style, size and content changes.
They also cover cases where we don&apos;t use the `WKSeparatedImageView` for
layers using tiling.

Canonical link: <a href="https://commits.webkit.org/288011@main">https://commits.webkit.org/288011@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/173b4a458f97168ab50f74478b6d22c930bc6010

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35613 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86203 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32654 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83770 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9008 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63708 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21434 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84734 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/872 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74315 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44000 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/772 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28492 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31109 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29089 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87643 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8901 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6304 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72044 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9084 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70140 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71279 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17750 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15359 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14273 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8853 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14381 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8693 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12214 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10502 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp [hunk 1]
```diff
@@ -1313,6 +1313,19 @@ std::optional<PlatformLayerIdentifier> GraphicsLayerCA::contentsLayerIDForModel(
...
 #endif
 
+#if HAVE(CORE_ANIMATION_SEPARATED_LAYERS)
+void GraphicsLayerCA::setIsSeparatedImage(bool isSeparatedImage)
+{
+    if (m_isSeparatedImage == isSeparatedImage)
+        return;
+    m_isSeparatedImage = isSeparatedImage;
+    if (m_isSeparatedImage)
+        changeLayerTypeTo(PlatformCALayer::LayerType::LayerTypeSeparatedImageLayer);
```
Comment (smfr) [inline Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: You shouldn't call ` changeLayerTypeTo` here; this should only be called in the "commit" phase. See `commitLayerTypeChangesBeforeSublayers()`. Also note that you need to handled huge layers, so changes between separated and tiled.

You should also add some tests for toggling the state in both directions, including on huge layers.

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp [hunk 2]
```diff
@@ -4620,6 +4633,7 @@ void GraphicsLayerCA::changeLayerTypeTo(PlatformCALayer::LayerType newLayerType)
         backing->setTileCoverage(m_tileCoverage);
 
     m_layer->adoptSublayers(*oldLayer);
+    m_layer->setCornerRadius(oldLayer->cornerRadius());
```
Comment (smfr) [inline Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: We never read state off the old layer. `cornerRadius` needs to come from our internal state.

Is this change testable?

File: Source/WebCore/platform/graphics/GraphicsLayer.h [hunk 3]
```diff
@@ -568,9 +568,14 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     enum class ModelInteraction : uint8_t { Enabled, Disabled };
     virtual void setContentsToModel(RefPtr<Model>&&, ModelInteraction) { }
     virtual std::optional<PlatformLayerIdentifier> contentsLayerIDForModel() const { return std::nullopt; }
+#endif
+#if HAVE(CORE_ANIMATION_SEPARATED_LAYERS)
+    virtual bool isSeparatedImage() const { return false; }
```
Comment (smfr) [inline Source/WebCore/platform/graphics/GraphicsLayer.h]: At this level, do we care that it's a separated image, or just separated? I feel like the "image" part is more about layer contents.

File: Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h [hunk 4]
```diff
@@ -222,7 +223,7 @@ class RemoteLayerBackingStoreProperties {
     RemoteLayerBackingStoreProperties() = default;
     RemoteLayerBackingStoreProperties(RemoteLayerBackingStoreProperties&&) = default;
 
-    void applyBackingStoreToLayer(CALayer *, LayerContentsType, std::optional<WebCore::RenderingResourceIdentifier>, bool replayDynamicContentScalingDisplayListsIntoBackingStore);
+    void applyBackingStoreToLayer(CALayer *, LayerContentsType, std::optional<WebCore::RenderingResourceIdentifier>, bool replayDynamicContentScalingDisplayListsIntoBackingStore, UIView *);
```
Comment (smfr) [inline Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h]: The last parameter needs a name.

File: Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm [hunk 5]
```diff
@@ -501,13 +508,27 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
...
     layer.contentsOpaque = m_isOpaque;
 
+#if HAVE(CORE_ANIMATION_SEPARATED_LAYERS)
+    if (hostingView && [hostingView isKindOfClass:[WKSeparatedImageView class]] && contentsType == LayerContentsType::CachedIOSurface) {
+        auto machSendRight = std::get<MachSendRight>(WTFMove(*m_bufferHandle));
+        auto surface = WebCore::IOSurface::createFromSendRight(WTFMove(machSendRight));
+        if (surface) {
+            auto context = surface->createPlatformContext();
+            auto image = surface->createImage(context.get());
+            [(WKSeparatedImageView *)hostingView setCGImage:image.get() cgContext:context.get()];
```
Comment (smfr) [inline Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm]: Is it OK to pass in the CGContext from the IOSurface? What does the view do with it?

File: Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm [hunk 6]
```diff
@@ -262,14 +265,27 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
         auto* backingStore = properties.backingStoreOrProperties.properties.get();
         if (backingStore && properties.backingStoreAttached) {
             std::optional<WebCore::RenderingResourceIdentifier> asyncContentsIdentifier;
+            UIView* hostingView = nullptr;
```
Comment (smfr) [inline Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm]: `nil`

File: Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm [hunk 7]
```diff
@@ -262,14 +265,27 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
...
+#endif
             }
 
-            backingStore->applyBackingStoreToLayer(layer, layerContentsType, asyncContentsIdentifier, layerTreeHost->replayDynamicContentScalingDisplayListsIntoBackingStore());
-        } else
+            backingStore->applyBackingStoreToLayer(layer, layerContentsType, asyncContentsIdentifier, layerTreeHost->replayDynamicContentScalingDisplayListsIntoBackingStore(), hostingView);
+        } else {
+#if HAVE(CORE_ANIMATION_SEPARATED_LAYERS)
+            UIView *hostingView = layerTreeNode->uiView();
+            if (hostingView && [hostingView isKindOfClass:[WKSeparatedImageView class]])
```
Comment (smfr) [inline Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm]: It's not great that we pay the cost of `isKindOfClass:` for every single layer, when 99.99% of them will never be separated.

File: Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm (hunk 5 above)
Comment (etiennesegonzac) [reply Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm]: Just holds it, but maybe I misunderstood the comment in:
https://github.com/WebKit/WebKit/blob/756e76506075a07f145e541a5345bef05b432e3f/Source/WebCore/platform/graphics/cocoa/IOSurface.h#L149-L152

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp (hunk 2 above)
Comment (etiennesegonzac) [reply Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: yes!

File: Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm (hunk 7 above)
Comment (etiennesegonzac) [reply Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm]: doing the clean-up in the view subclass instead now

**Review by smfr — APPROVED** [review]

**Pull request merged**